{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Mutations\n",
    "\n",
    "### Problem\n",
    "\n",
    "To test our folding methods on different BM3 mutants, we'd replicate the mutations in the target sequence in the starting sequence. The PyRosetta mutate tool requires the number of the amino acid (the numbering system is different in some poses from different BM3s) and the letter code amino acid to change it to.  The numbering system we use starts at ```MTIKEM...``` but not all the structures have these amino acids, so have a different numbering system. It would be useful to translate any BM3 numbering sytem to ours, so we can easily identify the position of mutations.\n",
    "\n",
    "\n",
    "### Psuedo code\n",
    "\n",
    "Not real code, but it's useful sometimes to make sketches like this. It's a simplified version of what I think a test iteration might look like. First, starting and target poses are made and their sequences are extracted. I think the sequences are taken directly from the structure, so might not be the complete biological sequence. Then we have to find the mutations between the two sequences, which I've been having trouble with. I'll put some things that I tried below. \n",
    "\n",
    "```python\n",
    "starting_pose = pose(BM3wt.pdb)\n",
    "starting_sequence = get_seq(starting_strucutre)\n",
    "\n",
    "target_pose = pose(BM3mutant.pdb)\n",
    "target_sequence = get_seq(target_structure) # not in sync with starting_sequence\n",
    "\n",
    "mutations = find_mutations(starting_sequence, target_sequence)\n",
    "# the pyrosetta mutation tool requires the amino acid number\n",
    "# and the residue to change to\n",
    "\n",
    "mutant_pose = copy(starting_pose)\n",
    "mutant_pose.mutate(mutations) # something like that\n",
    "\n",
    "mutant_pose.fold()\n",
    "\n",
    "score(mutant_pose, target_pose)\n",
    "```\n",
    "\n",
    "\n",
    "### Aim\n",
    "* To make a function that reliably identifies mutations between two BM3 sequences\n",
    "* To return the mutation(s), where the numbering system is adjusted as necesary\n",
    "\n",
    "I've dumped in some functions that I was trying yesterday. The sequences of our (current) mutant group are at ```BM3-Design-PyRosetta/data/sequences/```, so I made some pandas functions to get them. I also looked at using alignments, but I've no idea if that's a good idea or a waste of time. Maybe it would be simpler to just find some conserved residues and search for them to find our sequence offset? Or maybe finding the mutations directly from the sequence isn't a good use of our time and we'd be better off just finding the mutations in literature?\n",
    "\n",
    "\n",
    "### Pandas functions\n",
    "These ones are probably fine, they read in fasta data as a pandas Series, or DataFrame. The DataFrame is good for alignments, because each aa gets a cell so we can use padnas operations to find conservation etc/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3ekb_clean.pdb    TIKEMPQPKTFGELKNLPLLNTDKPVQALMKIADELGEIFKFEAPG...\n",
       "2nnb_clean.pdb    KEMPQPKTFGELKNLPLLNTDKPVQALMKIADELGEIFKFEAPGRV...\n",
       "3ben_clean.pdb    EMPQPKTFGELKNLPLLNTDKPVQALMKIADELGEIFKFEAPGRVT...\n",
       "1yqo_clean.pdb    KEMPQPKTFGELKNLPLLNTDKPVQALMKIADELGEIFKFEAPGRV...\n",
       "1p0v_clean.pdb    KEMPQPKTFGELKNLPLLNTDKPVQALMKIADELGEIFKFEAPGRV...\n",
       "dtype: object"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "def FastaToSeries(path):\n",
    "    # opens fasta files, outputs series\n",
    "    # full fasta string in one cell\n",
    "    with open(path,'r') as file:\n",
    "        data = file.read()\n",
    "    data = [i.split('\\n') for i in data.split('>')]\n",
    "    index = [i[0] for i in data][1:] # first item is ''\n",
    "    values = [''.join(i[1:]) for i in data][1:] # first item is []\n",
    "    df = pd.Series(values, index=index)\n",
    "    return df\n",
    "\n",
    "FastaToSeries('../data/sequences/Sequences.fasta').head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>...</th>\n",
       "      <th>469</th>\n",
       "      <th>470</th>\n",
       "      <th>471</th>\n",
       "      <th>472</th>\n",
       "      <th>473</th>\n",
       "      <th>474</th>\n",
       "      <th>475</th>\n",
       "      <th>476</th>\n",
       "      <th>477</th>\n",
       "      <th>478</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4dqk_clean.pdb</th>\n",
       "      <td>M</td>\n",
       "      <td>H</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>F</td>\n",
       "      <td>S</td>\n",
       "      <td>T</td>\n",
       "      <td>N</td>\n",
       "      <td>V</td>\n",
       "      <td>V</td>\n",
       "      <td>...</td>\n",
       "      <td>A</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4dql_clean.pdb</th>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>F</td>\n",
       "      <td>S</td>\n",
       "      <td>T</td>\n",
       "      <td>N</td>\n",
       "      <td>V</td>\n",
       "      <td>V</td>\n",
       "      <td>...</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3qi8_clean.pdb</th>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>K</td>\n",
       "      <td>K</td>\n",
       "      <td>I</td>\n",
       "      <td>P</td>\n",
       "      <td>L</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>I</td>\n",
       "      <td>P</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3cbd_clean.pdb</th>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>K</td>\n",
       "      <td>K</td>\n",
       "      <td>I</td>\n",
       "      <td>P</td>\n",
       "      <td>L</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3psx_clean.pdb</th>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>K</td>\n",
       "      <td>K</td>\n",
       "      <td>I</td>\n",
       "      <td>P</td>\n",
       "      <td>L</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 479 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               0   1   2   3   4   5   6   7   8   9    ... 469 470 471 472  \\\n",
       "4dqk_clean.pdb   M   H   G   A   F   S   T   N   V   V  ...   A   -   -   -   \n",
       "4dql_clean.pdb   -   -   G   A   F   S   T   N   V   V  ...   A   G   -   -   \n",
       "3qi8_clean.pdb   -   -   -   -   -   -   -   -   -   -  ...   K   K   I   P   \n",
       "3cbd_clean.pdb   -   -   -   -   -   -   -   -   -   -  ...   K   K   I   P   \n",
       "3psx_clean.pdb   -   -   -   -   -   -   -   -   -   -  ...   K   K   I   P   \n",
       "\n",
       "               473 474 475 476 477 478  \n",
       "4dqk_clean.pdb   -   -   -   -   -   -  \n",
       "4dql_clean.pdb   -   -   -   -   -   -  \n",
       "3qi8_clean.pdb   L   G   G   I   P   S  \n",
       "3cbd_clean.pdb   L   -   -   -   -   -  \n",
       "3psx_clean.pdb   L   -   -   -   -   -  \n",
       "\n",
       "[5 rows x 479 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def FastaToDataFrame(path):\n",
    "    # opens fasta files, outputs dataframe\n",
    "    # good for aligned sequences\n",
    "    with open(path,'r') as file:\n",
    "        data = file.read()\n",
    "    data = [i.split('\\n') for i in data.split('>')]\n",
    "    index = [i[0] for i in data][1:] # first item is ''\n",
    "    values = [list(''.join(i[1:])) for i in data][1:] # first item is []\n",
    "    df = pd.DataFrame(values, index=index)\n",
    "    return df\n",
    "\n",
    "\n",
    "\n",
    "FastaToDataFrame('../data/sequences/Sequences_msa.fasta').head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Alingments\n",
    "Still not sure this is a good idea! We can put the aligned sequences into a DataFrame  with\n",
    "```python\n",
    "pd.DataFrame([list(seq1), list(seq2)])\n",
    "```\n",
    "\n",
    "and we see that the mutations are offset from eachother, maybe this is solvable??"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'aln_reference_seq': 'TIKEMPQPKTFGELKNLPLLNTDKPVQALMKIADELGEIFKFEAPGRVTRYLSSQRLIKEACDESRFDKNLSQALKFVRDFAGDGLFTSWTHEKNWKKAHNILLPSFSQQAMKGYHAMMVDIAVQLVQKWERLNADEHIEVPEDMTRLTLDTIGLCGFNYRFNSFYRDQPHPFITSMVRALDEAMNKLQRAN--QFQEDIKVMNDLVDKIIADRKA---QSDDLLTHMLNGKDPETGEPLDDENIRYQIITFLIC-GHETTSGLLSFALYFLVKNPHVLQKAAEEAARVLVDPVPSYKQVKQLKYVGMVLNEALRLWPTAPAFSLYAKEDTVLGGEYPLEKGDELMVLIPQLHRDKTIWGDDVEEFRPERFENPSAIPQHAFKPFGNGQRACIGQ-QFALHEATLVLGMMLKHFDFEDHTNYELDIKETLTLKPEGFVVKAKSKKIPLG',\n",
       " 'aln_query_seq': '--KEMPQPKTFGELKNLPLLNTDKPVQALMKIADELGEIFKFEAPGRVTRYLSSQRLIKEACDESRFDKNLSQALKFVRDFAGDGLFTSWTHEKNWKKAHNILLPSFSQQAMKGYHAMMVDIAVQLVQKWERLNADEHIEVPEDMTRLTLDTIGLCGFNYRFNSFYRDQPHPFITSMVRALDEAMNKL---NKRQFQEDIKVMNDLVDKIIADRKASGEQSDDLLTHMLNGKDPETGEPLDDENIRYQIITFLI-AGHETTSGLLSFALYFLVKNPHVLQKAAEEAARVLVDPVPSYKQVKQLKYVGMVLNEALRLWPTAPAFSLYAKEDTVLGGEYPLEKGDELMVLIPQLHRDKTIWGDDVEEFRPERFENPSAIPQHAFKPFGNGQRACIG-KQFALHEATLVLGMMLKHFDFEDHTNYELDIKETLTLKPEGFVVKAKSKKIPL-',\n",
       " 'aln_score': 434.0}"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from Bio import pairwise2\n",
    "\n",
    "sequences = FastaToSeries('../data/sequences/Sequences.fasta')\n",
    "\n",
    "\n",
    "def AlignAgainstReferenceSequence(reference_seq, query_seq):\n",
    "    alignments = pairwise2.align.globalxx(reference_seq, query_seq) # makes several alignments\n",
    "    all_scores = [i[2] for i in alignments] # alignments return a list of lists, unpacking items\n",
    "    top_scores_indexes = [i for i,j in enumerate(all_scores) if j ==max(all_scores)]# list of indexes where score is max()\n",
    "    top_alignment = alignments[top_scores_indexes[0]] # Highest scoring first\n",
    "\n",
    "    dictionary = {'aln_reference_seq': top_alignment[0],\n",
    "    'aln_query_seq':top_alignment[1],\n",
    "        'aln_score': top_alignment[2]}\n",
    "    return dictionary\n",
    "\n",
    "AlignAgainstReferenceSequence(sequences[0], sequences[1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.9665924276169265"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "alnmnt = AlignAgainstReferenceSequence(sequences[0], sequences[1])\n",
    "\n",
    "def ResidueConservation(seq1, seq2):\n",
    "    ## takes two aligned sequences\n",
    "    # probably won't work if  they're different lengths\n",
    "    alignment_df = pd.DataFrame([list(seq1), list(seq2)])\n",
    "    alignment_df.replace(' ','-', inplace = True)\n",
    "    conserved_residue_count = 0\n",
    "    for i in alignment_df:\n",
    "        if len(alignment_df[i].unique()) <2:\n",
    "            conserved_residue_count += 1\n",
    "    frac_conserved = conserved_residue_count/len(seq1)\n",
    "    return frac_conserved\n",
    "\n",
    "\n",
    "ResidueConservation(alnmnt['aln_reference_seq'], alnmnt['aln_query_seq'])"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
